File: 2026-1/sesion-de-clase-07-08/SoftProg/SoftProgPersistencia/src/main/java/pe/edu/pucp/softprog/dao/clientes/ClienteDAOImpl.java
package pe.edu.pucp.softprog.dao.clientes;

import pe.edu.pucp.softprog.db.DBFactoryProvider;
import pe.edu.pucp.softprog.db.DBManager;
import pe.edu.pucp.softprog.dao.cuentas.CuentaUsuarioDAOImpl;
import pe.edu.pucp.softprog.modelo.Genero;
import pe.edu.pucp.softprog.modelo.clientes.CategoriaCliente;
import pe.edu.pucp.softprog.modelo.clientes.Cliente;

import java.sql.Connection;
import java.sql.Date;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;
import java.util.ArrayList;
import java.util.List;

public class ClienteDAOImpl implements ClienteDAO {
    @Override
    public Integer crear(Cliente modelo) {
        DBManager dbManager = DBFactoryProvider.getManager();
        try (Connection conn = dbManager.getConnection();
             PreparedStatement cmd = this.comandoCrear(conn, modelo)) {
            if (cmd.executeUpdate() == 0) {
                return null;
            }
            try (ResultSet rs = cmd.getGeneratedKeys()) {
                return rs.next() ? rs.getInt(1) : null;
            }
        }
        catch (SQLException e) {
            System.err.println("Error SQL: " + e.getMessage());
            throw new RuntimeException(e);
        }
        catch (Exception e) {
            System.err.println("Error inesperado: " + e.getMessage());
            throw new RuntimeException(e);
        }
    }

    @Override
    public boolean actualizar(Cliente modelo) {
        DBManager dbManager = DBFactoryProvider.getManager();
        try (Connection conn = dbManager.getConnection();
             PreparedStatement cmd = this.comandoActualizar(conn, modelo)) {
            return cmd.executeUpdate() > 0;
        }
        catch (SQLException e) {
            System.err.println("Error SQL: " + e.getMessage());
            throw new RuntimeException(e);
        }
        catch (Exception e) {
            System.err.println("Error inesperado: " + e.getMessage());
            throw new RuntimeException(e);
        }
    }

    @Override
    public boolean eliminar(int id) {
        DBManager dbManager = DBFactoryProvider.getManager();
        try (Connection conn = dbManager.getConnection();
             PreparedStatement cmd = this.comandoEliminar(conn, id)) {
            return cmd.executeUpdate() > 0;
        }
        catch (SQLException e) {
            System.err.println("Error SQL: " + e.getMessage());
            throw new RuntimeException(e);
        }
        catch (Exception e) {
            System.err.println("Error inesperado: " + e.getMessage());
            throw new RuntimeException(e);
        }
    }

    @Override
    public Cliente leer(int id) {
        DBManager dbManager = DBFactoryProvider.getManager();
        try (Connection conn = dbManager.getConnection();
             PreparedStatement cmd = this.comandoLeer(conn, id)) {
            ResultSet rs = cmd.executeQuery();
            if (!rs.next()) {
                System.err.println("No se encontro el registro con id: " + id);
                return null;
            }
            return this.mapearModelo(rs);
        }
        catch (SQLException e) {
            System.err.println("Error SQL: " + e.getMessage());
            throw new RuntimeException(e);
        }
        catch (Exception e) {
            System.err.println("Error inesperado: " + e.getMessage());
            throw new RuntimeException(e);
        }
    }

    @Override
    public List<Cliente> leerTodos() {
        DBManager dbManager = DBFactoryProvider.getManager();
        try (Connection conn = dbManager.getConnection();
             PreparedStatement cmd = this.comandoLeerTodos(conn)) {
            ResultSet rs = cmd.executeQuery();
            List<Cliente> modelos = new ArrayList<>();
            while (rs.next()) {
                modelos.add(this.mapearModelo(rs));
            }
            return modelos;
        }
        catch (SQLException e) {
            System.err.println("Error SQL: " + e.getMessage());
            throw new RuntimeException(e);
        }
        catch (Exception e) {
            System.err.println("Error inesperado: " + e.getMessage());
            throw new RuntimeException(e);
        }
    }

    @Override
    public Cliente buscarPorDni(String dni) {
        DBManager dbManager = DBFactoryProvider.getManager();
        try (Connection conn = dbManager.getConnection();
             PreparedStatement cmd = this.comandoBuscarPorDni(conn, dni)) {
            ResultSet rs = cmd.executeQuery();
            return rs.next() ? this.mapearModelo(rs) : null;
        }
        catch (SQLException e) {
            System.err.println("Error SQL: " + e.getMessage());
            throw new RuntimeException(e);
        }
        catch (Exception e) {
            System.err.println("Error inesperado: " + e.getMessage());
            throw new RuntimeException(e);
        }
    }

    protected PreparedStatement comandoCrear(Connection conn,
                                             Cliente modelo) throws SQLException {
        String sql = """
            INSERT INTO CLIENTE (
                idCuentaUsuario,
                dni,
                nombre,
                apellidoPaterno,
                genero,
                fechaNacimiento,
                categoria,
                lineaCredito,
                activo
            ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)
            """;
        PreparedStatement cmd = conn.prepareStatement(sql, Statement.RETURN_GENERATED_KEYS);
        if (modelo.getCuentaUsuario() == null) {
            cmd.setNull(1, java.sql.Types.INTEGER);
        }
        else {
            cmd.setInt(1, modelo.getCuentaUsuario().getId());
        }
        cmd.setString(2, modelo.getDni());
        cmd.setString(3, modelo.getNombre());
        cmd.setString(4, modelo.getApellidoPaterno());
        cmd.setString(5, modelo.getGenero().name());
        cmd.setDate(6, new Date(modelo.getFechaNacimiento().getTime()));
        cmd.setString(7, modelo.getCategoria().name());
        cmd.setDouble(8, modelo.getLineaCredito());
        cmd.setBoolean(9, modelo.isActivo());
        return cmd;
    }

    protected PreparedStatement comandoActualizar(Connection conn,
                                                  Cliente modelo) throws SQLException {
        String sql = """
            UPDATE CLIENTE
            SET idCuentaUsuario = ?,
                dni = ?,
                nombre = ?,
                apellidoPaterno = ?,
                genero = ?,
                fechaNacimiento = ?,
                categoria = ?,
                lineaCredito = ?,
                activo = ?
            WHERE id = ?
            """;
        PreparedStatement cmd = conn.prepareStatement(sql);
        if (modelo.getCuentaUsuario() == null) {
            cmd.setNull(1, java.sql.Types.INTEGER);
        }
        else {
            cmd.setInt(1, modelo.getCuentaUsuario().getId());
        }
        cmd.setString(2, modelo.getDni());
        cmd.setString(3, modelo.getNombre());
        cmd.setString(4, modelo.getApellidoPaterno());
        cmd.setString(5, modelo.getGenero().name());
        cmd.setDate(6, new Date(modelo.getFechaNacimiento().getTime()));
        cmd.setString(7, modelo.getCategoria().name());
        cmd.setDouble(8, modelo.getLineaCredito());
        cmd.setBoolean(9, modelo.isActivo());
        cmd.setInt(10, modelo.getId());
        return cmd;
    }

    protected PreparedStatement comandoEliminar(Connection conn,
                                                Integer id) throws SQLException {
        String sql = "DELETE FROM CLIENTE WHERE id = ?";
        PreparedStatement cmd = conn.prepareStatement(sql);
        cmd.setInt(1, id);
        return cmd;
    }

    protected PreparedStatement comandoLeer(Connection conn,
                                            Integer id) throws SQLException {
        String sql = "SELECT * FROM CLIENTE WHERE id = ?";
        PreparedStatement cmd = conn.prepareStatement(sql);
        cmd.setInt(1, id);
        return cmd;
    }

    protected PreparedStatement comandoLeerTodos(Connection conn) throws SQLException {
        String sql = "SELECT * FROM CLIENTE";
        return conn.prepareStatement(sql);
    }

    protected PreparedStatement comandoBuscarPorDni(Connection conn,
                                                    String dni) throws SQLException {
        String sql = "SELECT * FROM CLIENTE WHERE dni = ?";
        PreparedStatement cmd = conn.prepareStatement(sql);
        cmd.setString(1, dni);
        return cmd;
    }

    protected Cliente mapearModelo(ResultSet rs) throws SQLException {
        Cliente modelo = new Cliente();
        modelo.setId(rs.getInt("id"));

        int idCuentaUsuario = rs.getInt("idCuentaUsuario");
        if (!rs.wasNull()) {
            modelo.setCuentaUsuario(new CuentaUsuarioDAOImpl().leer(idCuentaUsuario));
        }

        modelo.setDni(rs.getString("dni"));
        modelo.setNombre(rs.getString("nombre"));
        modelo.setApellidoPaterno(rs.getString("apellidoPaterno"));
        modelo.setGenero(Genero.valueOf(rs.getString("genero")));
        modelo.setFechaNacimiento(rs.getDate("fechaNacimiento"));
        modelo.setCategoria(CategoriaCliente.valueOf(rs.getString("categoria")));
        modelo.setLineaCredito(rs.getDouble("lineaCredito"));
        modelo.setActivo(rs.getBoolean("activo"));
        return modelo;
    }
}

File: 2026-1/sesion-de-clase-12/con-transacciones-sql/SoftProgPersistencia/Dao/Clientes/ClienteDaoImpl.cs
using System.Data;
using System.Data.Common;
using SoftProgModelo.Modelos.Clientes;
using SoftProgPersistencia.Dao.Cuentas;

namespace SoftProgPersistencia.Dao.Clientes;

public class ClienteDaoImpl : PersonaBaseDao<Cliente>, IClienteDao
{
    protected override DbCommand ComandoCrear(DbConnection conn, Cliente modelo)
    {
        var cmd = conn.CreateCommand();
        cmd.CommandText = @"INSERT INTO CLIENTE (idCuentaUsuario, dni, nombre, apellidoPaterno, genero, fechaNacimiento, categoria, lineaCredito, activo)
                            VALUES (@p1, @p2, @p3, @p4, @p5, @p6, @p7, @p8, @p9);";
        SetCamposCliente(cmd, modelo);
        return cmd;
    }

    protected override DbCommand ComandoActualizar(DbConnection conn, Cliente modelo)
    {
        var cmd = conn.CreateCommand();
        cmd.CommandText = @"UPDATE CLIENTE
                            SET idCuentaUsuario = @p1, dni = @p2, nombre = @p3, apellidoPaterno = @p4,
                                genero = @p5, fechaNacimiento = @p6, categoria = @p7, lineaCredito = @p8, activo = @p9
                            WHERE id = @p10;";
        SetCamposCliente(cmd, modelo);
        CrearParametro(cmd, "@p10", modelo.Id);
        return cmd;
    }

    protected override DbCommand ComandoEliminar(DbConnection conn, int id)
    {
        var cmd = conn.CreateCommand();
        cmd.CommandText = "DELETE FROM CLIENTE WHERE id = @p1;";
        CrearParametro(cmd, "@p1", id);
        return cmd;
    }

    protected override DbCommand ComandoLeer(DbConnection conn, int id)
    {
        var cmd = conn.CreateCommand();
        cmd.CommandText = "SELECT * FROM CLIENTE WHERE id = @p1;";
        CrearParametro(cmd, "@p1", id);
        return cmd;
    }

    protected override DbCommand ComandoLeerTodos(DbConnection conn)
    {
        var cmd = conn.CreateCommand();
        cmd.CommandText = "SELECT * FROM CLIENTE;";
        return cmd;
    }

    protected override DbCommand ComandoBuscarPorDni(DbConnection conn, string dni)
    {
        var cmd = conn.CreateCommand();
        cmd.CommandText = "SELECT * FROM CLIENTE WHERE dni = @p1;";
        CrearParametro(cmd, "@p1", dni);
        return cmd;
    }

    protected override Cliente MapearModelo(DbDataReader reader)
    {
        var modelo = new Cliente
        {
            Id = LeerEntero(reader, "id"),
            Categoria = Enum.Parse<CategoriaCliente>(LeerTexto(reader, "categoria")),
            LineaCredito = LeerDecimal(reader, "lineaCredito"),
            Activo = LeerBooleano(reader, "activo")
        };

        var idCuentaUsuario = LeerEnteroNullable(reader, "idCuentaUsuario");
        if (idCuentaUsuario.HasValue)
        {
            modelo.CuentaUsuario = new CuentaUsuarioDaoImpl().Leer(idCuentaUsuario.Value);
        }

        MapearCamposPersona(reader, modelo);
        return modelo;
    }

    private void SetCamposCliente(DbCommand cmd, Cliente modelo)
    {
        CrearParametro(cmd, "@p1", GetIdCuentaUsuario(modelo), DbType.Int32);
        CrearParametro(cmd, "@p2", modelo.Dni);
        CrearParametro(cmd, "@p3", modelo.Nombre);
        CrearParametro(cmd, "@p4", modelo.ApellidoPaterno);
        CrearParametro(cmd, "@p5", modelo.Genero.ToString());
        CrearParametro(cmd, "@p6", modelo.FechaNacimiento, DbType.Date);
        CrearParametro(cmd, "@p7", modelo.Categoria.ToString());
        CrearParametro(cmd, "@p8", modelo.LineaCredito);
        CrearParametro(cmd, "@p9", modelo.Activo);
    }
}
